tools/system_libs.py

```diff
@@ -281,6 +292,100 @@ def get_files(self):
 
     raise NotImplementedError()
 
+  def write_ninja_file(self, filename, libname):
+    cflags = self.get_cflags()
+    asflags = get_base_cflags()
+    # TODO(sbc) There is an llvm bug that causes a crash when `-g` is used with
+    # assembly files that define wasm globals.
+    asflags = [arg for arg in asflags if arg != '-g']
+    cflags_asm = [arg for arg in cflags if arg != '-g']
+
+    def join(flags):
+      return ' '.join(flags)
+
+    out = f'''\
+# Automatically generated by tools/system_libs.py.  DO NOT EDIT
+
+ninja_required_version = 1.5
+
+ASFLAGS = {join(asflags)}
+CFLAGS = {join(cflags)}
+CFLAGS_ASM = {join(cflags_asm)}
+EMCC = {shared.EMCC}
+EMXX = {shared.EMXX}
+EMAR = {shared.EMAR}
+
+rule cc
+  depfile = $out.d
+  command = $EMCC -MD -MF $out.d $CFLAGS -c $in -o $out
+  description = CC $in
+
+rule cxx
+  depfile = $out.d
+  command = $EMXX -MD -MF $out.d $CFLAGS -c $in -o $out
+  description = CXX $out
+
+rule asm
+  command = $EMCC $ASFLAGS -c $in -o $out
+  description = ASM $out
+
+rule asm_cpp
+  depfile = $out.d
+  command = $EMCC -MD -MF $out.d $CFLAGS_ASM -c $in -o $out
+  description = ASM $out
+
+rule direct_cc
+  depfile = $with_depfile
+  command = $EMCC -MD -MF $with_depfile $CFLAGS -c $in -o $out
+  description = CC $out
+
+rule archive
+  command = $EMAR cr $out $in
+  description = AR $out
+
+'''
+    suffix = shared.suffix(libname)
+    input_files = self.get_files()
+
+    if suffix == '.o':
+      assert len(input_files) == 1
+      depfile = shared.unsuffixed_basename(input_files[0]) + '.d'
+      out += f'build {libname}: direct_cc {input_files[0]}\n'
+      out += f'  with_depfile = {depfile}\n'
+    else:
+      objects = []
+      for src in input_files:
+        o = shared.unsuffixed_basename(src) + '.o'
```

Comment:
Oops, I wrote a prototype of this and I accidentally just pushed it onto this branch instead of a separate branch. Feel free to force-push over it if you want.

But the results are very promising. I tested building the emscripten system libraries without ninja, with one-lib-at-a-time ninja as in your branch, and all-at-once ninja in my diff, using 8 cores on my workstation:
Current: 11:05 (665s)
Ninja: 10:00 (600s)
Ninja w/deferred build: 8:10 (490s)
So 26% faster.

Using all 72 cores, the difference is huge, 314s vs 96s

Reply:
I also go this working in my branch, and created a master ninja file but I didn't love the ergonomics.   I'd like to land this change first and then consider that option separately.     This patch solve the problem that a lot of us run into which is that we currently need to rebuilt a lot of libraries completely when we change just a few files. 

The master-ninja things really only helps when using embuilder to build everything at once.. which may be common, but also should be lass common once we have dependencies in place.

Reply:
Yes, doing this first independently makes sense.